[MLIR][Tensor] Enhance bufferization of tensor.expand_shape op

arnab-polymage opened the pull request:
Instead of inferring the output shape argument of
memref.expand_shape op, use output_shape argument of tensor.expand_shape op by adding dynamic dimension support for bufferization of tensor.expand_shape when there are more than one dynamic dim within a reassociation set.

arnab-polymage commented:
@ramiro050  @bondhugula please review

MaheshRavishankar commented on mlir/lib/Dialect/Tensor/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -337,14 +337,27 @@ struct ExpandShapeOpInterface
-        expandShapeOp.getReassociationIndices());
+    // Use output_shape argument of tensor.expand_shape op to get the result
+    // shapes of the memref.expand_shape op to be created.
+    SmallVector<OpFoldResult> outShape;
```
I think you should be able to use `expandShapeOp.getMixedOutputShape()`.

MaheshRavishankar commented on mlir/lib/Dialect/Tensor/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -337,14 +337,27 @@ struct ExpandShapeOpInterface
+    auto memrefExpandShape = rewriter.create<memref::ExpandShapeOp>(
+        op->getLoc(), tensorResultType.getShape(), *buffer,
+        expandShapeOp.getReassociationIndices(), outShape);
+    SmallVector<int64_t> staticShape;
```
You can use `dispatchIndexOpFoldResult`.

hanhanW commented:
> Instead of inferring the output shape argument of memref.expand_shape op, use output_shape argument of tensor.expand_shape op by adding dd support for bufferization of tensor.expand_shape when there are more than one dynamic dim within a reassociation set.

What is dd support?

MaheshRavishankar commented:
> > Instead of inferring the output shape argument of memref.expand_shape op, use output_shape argument of tensor.expand_shape op by adding dd support for bufferization of tensor.expand_shape when there are more than one dynamic dim within a reassociation set.
> 
> What is dd support?

I read it as dynamic dimensions :)

arnab-polymage committed:
[MLIR][Tensor] Enhance bufferization of tensor.expand_shape op

Instead of inferring the output shape argument of
memref.expand_shape op, use output_shape argument of
tensor.expand_shape op by adding dynamic dimension support
for bufferization of tensor.expand_shape when there are
more than one dynamic dim within a reassociation set.
